mixer/pkg/attribute/list.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package attribute
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// List wraps a list of values and optionally reference counts it
+type List struct {
+	// name of the list
+	name string
+	// entries in the list
+	entries []interface{}
+}
+
+// NewList creates a new list
+func NewList(name string) List {
+	return List{
+		name:    name,
+		entries: make([]interface{}, 0, 1),
+	}
+}
+
+// NewListForTesting should only be used for testing.
+func NewListForTesting(name string, entries []interface{}) List {
+	return List{
+		name:    name,
+		entries: entries,
+	}
+}
+
+// Append adds an element to the list
```

Comment:
This comment is a bit misleading. It's not appending to the list, it's producing a new list which is a concatenation of the old list with a new element.

Also, although this seem like a functional immutable copy, this is not so given the semantics of the underlying slice. If you have l1:

  l2 = l1.Append(x)
  l3 = l1.Append(x)

This will not work as you'd expect with a purely functional design. 

So it might be cleaner to just have Append take a (l*List) and mutate it in place without returning a List.

Reply:
Good point. I changed it to take a pointer and mutate in place.